rex/external/rexarray.py

```diff
@@ -0,0 +1,496 @@
+
+    def __init__(self, manager, group=None, mode=None, lock=HDF5_LOCK):
+        if isinstance(manager, h5py.File | h5py.Group):
+            if group is None:
+                root, group = find_root_and_group(manager)
+            else:
+                if type(manager) is not h5py.File:
+                    raise ValueError(
+                        "must supply a h5py.File if the group "
+                        "argument is provided"
+                    )
+                root = manager
+            manager = DummyFileManager(root)
+
+        self._manager = manager
+        self._group = group
+        self._mode = mode
+        self._filename = _get_h5_fn(self.ds)
+        self._ds_shape = None
+        self.lock = ensure_lock(lock)
+
+    @classmethod
+    def open(cls, filename, mode="r", group=None, lock=None, driver=None,
+             driver_kwds=None, storage_options=None):
+        """_summary_
+
+        Parameters
+        ----------
+        filename : _type_
+            _description_
+        mode : str, optional
+            _description_. By default, ``"r"``.
+        group : _type_, optional
+            _description_. By default, ``None``.
+        lock : _type_, optional
+            _description_. By default, ``None``.
+        driver : _type_, optional
+            _description_. By default, ``None``.
+        driver_kwds : _type_, optional
+            _description_. By default, ``None``.
+        storage_options : dict[str, Any] | None, optional
+            _description_. By default, ``None``.
+
+        Returns
+        -------
+        _type_
+            _description_
+
+        Raises
+        ------
+        ValueError
+            _description_
+        ValueError
+            _description_
+        """
+        remote_file = (isinstance(filename, str)
+                       and is_remote_uri(filename)
+                       and driver is None)
+        if remote_file:
+            mode_ = "rb" if mode == "r" else mode
+            filename = _open_remote_file(
+                filename, mode=mode_, storage_options=storage_options
+            )
+
+        if isinstance(filename, bytes):
+            raise ValueError("can't open netCDF4/HDF5 as bytes "
+                             "try passing a path or file-like object")
+        elif isinstance(filename, io.IOBase):
+            magic_number = read_magic_number_from_file(filename)
+            if not magic_number.startswith(b"\211HDF\r\n\032\n"):
+                raise ValueError(f"{magic_number!r} is not the signature "
+                                 "of a valid netCDF4 file")
+
+        kwargs = {"driver": driver}
+        if driver_kwds is not None:
+            kwargs.update(driver_kwds)
+
+        if lock is None:
+            if mode == "r":
+                lock = HDF5_LOCK
+            else:
+                lock = combine_locks([HDF5_LOCK, get_write_lock(filename)])
+
+        manager = CachingFileManager(h5py.File, filename, mode=mode,
```

Comment:
Okay so you might have to walk me through how all of this works, but is this where you define how the file is actually being opened? and with what object handler? 

My extra ask (or just naive question) is whether we can open an S3 or HSDS path with xarray.open_dataset() by using `h5pyd.File`? It seems like maybe we can open an S3 path but my kernel just hangs when i try. I can't tell if it's latency because its doing a lot behind the scenes or if S3/fsspec is just not supported.

Reply:
IDK if we _can_ support those but I definitely want to if possible. Adding/validating h5py/fsspec is still on my TODO list (if it works already, its basically by accident). I plan to add tests for these cases as well, assuming it's possible to support these cases out of the box. 

Sorry for dragging my feet on this code, I haven't touched it in a while. My plan is to get back into it this week

Reply:
No worries at all, i haven't had time to review or provide feedback! I'm basically trying to figure out where the h5 file is actually opened and where we specify which object opens it. I think if we can figure that out we could add hsds/fsspec integration? Why don't you reach out when you start work on this again and explain stuff to me and maybe i can help or brainstorm.

Reply:
So actually, the file is opened pretty much where you left this comment. 

Definitely still happy to walk though this code - just let me know when you have time!

As of now, HSDS and S3 are explicitly supported. I have added tests for reading files using xarray in both the hsds and s3 test files.

Reply:
Wow really nice work! It's really fast with hsds. So awesome!!!